### DISCUSSION TITLE
Negation

### DISCUSSION THREAD

**Discussion by voidpetal** [LukasZahradnik/PyNeuraLogic#40] [General]
Hello, 

I am trying to make a predicate `a_or_b` which fires iff all nodes are of types A or B. I was thinking that it may be easier to define a predicate `not_a_or_b` which fires only if there exists a node which is not of type A and not of type B, and then negate it to get the original predicate. I tried using `R.a_or_b <= -R.not_a_or_b` and `R.a_or_b <= ~R.not_a_or_b`, but both times it resolves into `reverse` predicate when I print the template. 
Am I using the right approach to negation here?

On the other hand, without using negation, I could write something like this:
```
R.a_or_b <= R.a(V.X)
R.a_or_b <= R.b(V.X)
```
But then it would not guarantee the nonexistence of a node of another type.

Thank you in advance for your advice!

**Comment by LukasZahradnik** [comment]
Hi,

There isn't currently a negation in PyNeuraLogic. There used to be one, but it was later fully replaced with the reverse function.

It was removed because it behaved (imo) in a weird way. I'm trying to re-introduce it, and it works fine in the InferenceEngine, but not in regular scenarios. I will investigate more.

I have no idea how to resolve your problem without negation other than manually adding `R.negated_a`, etc. to examples...

**Comment by LukasZahradnik** [comment]
Okay, I found out what the issue was - neuralization did not work for negated relations. The solution was to set negated relations as hidden automatically.

The negation is now available in the latest version (0.7.3), and relations are negated via the tilde (~) operator. Minus is still applying the reverse function.

Be careful how you will use it. Negated relations will not produce any value, as they are hidden. So unless you add another body atom to this rule, you can only use `R.a_or_b` as a hidden relation.
```
R.a_or_b <= ~R.not_a_or_b
```

**Comment by voidpetal** [comment]
Thank you for your reply! Can you please explain what does it mean for a relation to be hidden?

**Reply by LukasZahradnik** [reply]
You can read about the hidden modifier here https://pyneuralogic.readthedocs.io/en/latest/advanced/modifiers.html#hidden-modifier

Basically, when you have a rule like this:
```py
R.h(V.X) <= (R.f(V.X), R.edge(V.X, V.Y))
```
You will compute the `h(X) = value of f(X) + value of edge(X, Y)`.

When you set the edge relation as hidden, that is, via `R.hidden` or prepending the name with `_`, the computation is just `h(X) = f(X)`.
```py
R.h(V.X) <= (R.f(V.X), R.hidden.edge(V.X, V.Y))
# or
R.h(V.X) <= (R.f(V.X), R._edge(V.X, V.Y))
```

When it comes to negation, negated relations are automatically set as hidden (there is no value assigned to negated relations anyway).

Problems arise when you have only hidden relations in a rule's body because then you cannot get its value. So, in that case, the rule's head has to be also used only in the hidden context.

```py
R.neg_x(V.X) <= ~R.x(V.X)
R.h(V.X) <= (R.hidden.neg_x(V.X), R.features(V.X))  # Will work -> h(X) = features(X)
```

```py
R.neg_x(V.X) <= ~R.x(V.X)
R.h(V.X) <= (R.neg_x(V.X), R.features(V.X))  # Will not work - R.neg_x has no value
```

**Comment by GustikS** [comment]
feature reminder: using default values for negated literals in the next release might be a good idea...

**Reply by GustikS** [reply]
plus update to full support of logic programs with negation with an automated check for stratification: https://github.com/GustikS/NeuraLogic/commit/7c1f2940f994bb1f7e101207db946e80c3abef7e
